From ca16dfe52ae4ba557d1108ea8cebae4c1a5335bf Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius.vlad@collabora.com>
Date: Mon, 21 Jul 2025 20:52:50 +0300
Subject: [PATCH] compositor: Allow occluded surfaces to receive frame
 callbacks

With commit b2e6a6438f7, "libweston: Don't add frame callbacks from
occluded paint nodes" we made it so we do not send out frame callback if
the paint nodes are not visible. This has the side-effect of
affecting also paint nodes/views which were previously added into
hidden layers (by the shells).

This is a compromise to still allow clients still receive frame callbacks
whenever the shell decide that will move put that view into a hidden layer.

By default the normal behaviour will be kept (not sending frame
callbacks to occlduded views) and with this
'send-frame-callback-to-occluded-surface' will alter that behaviour.

Upstream-Status: Pending
Signed-off-by: Marius Vlad <marius.vlad@collabora.com>
---
 frontend/main.c               | 2 ++
 include/libweston/libweston.h | 3 +++
 libweston/compositor.c        | 5 ++++-
 3 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/frontend/main.c b/frontend/main.c
index d135e77..2339209 100644
--- a/frontend/main.c
+++ b/frontend/main.c
@@ -4640,6 +4640,8 @@ wet_main(int argc, char *argv[], const struct weston_testsuite_data *test_data)
 
 	weston_config_section_get_bool(section, "require-input",
 				       &wet.compositor->require_input, true);
+	weston_config_section_get_bool(section, "send-frame-callback-to-occluded-surface",
+				       &wet.compositor->send_frame_cb_occluded_surface, false);
 
 	wet.require_outputs = REQUIRE_OUTPUTS_ANY;
 	weston_config_section_get_string(section, "require-outputs",
diff --git a/include/libweston/libweston.h b/include/libweston/libweston.h
index 4d3930d..947b921 100644
--- a/include/libweston/libweston.h
+++ b/include/libweston/libweston.h
@@ -1592,6 +1592,9 @@ struct weston_compositor {
 	/* Whether to let the compositor run without any input device. */
 	bool require_input;
 
+	/* Whether to send frame callbacks to occluded surfaces */
+	bool send_frame_cb_occluded_surface;
+
 	/* Whether to load multiple backends. */
 	bool multi_backend;
 
diff --git a/libweston/compositor.c b/libweston/compositor.c
index 84c25b6..288e67e 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -3746,6 +3746,8 @@ weston_output_repaint(struct weston_output *output, struct timespec *now)
 	int r;
 	uint32_t frame_time_msec;
 	enum weston_hdcp_protection highest_requested = WESTON_HDCP_DISABLE;
+	bool send_frame_cb_occluded_surface =
+		ec->send_frame_cb_occluded_surface;
 
 	TL_POINT(ec, "core_repaint_begin", TLP_OUTPUT(output), TLP_END);
 
@@ -3822,7 +3824,8 @@ weston_output_repaint(struct weston_output *output, struct timespec *now)
 		 * feedback to the respective lists if pnode/surface is
 		 * occluded
 		 */
-		if (!pixman_region32_not_empty(&pnode->visible))
+		if (!pixman_region32_not_empty(&pnode->visible) &&
+		    !send_frame_cb_occluded_surface)
 			continue;
 
 		wl_list_insert_list(&frame_callback_list,
-- 
2.47.2

